kale/embed/gripnet.py

```diff
@@ -241,3 +245,294 @@ def forward(
+
+        self.__check_supergraph__()
+        self.__init_supervertex_module_dict__()
+
+    def __check_supergraph__(self) -> None:
+        """Check if the input supergraph has parameter settings"""
+
+        if self.supergraph.supervertex_setting_dict is None:
+            error_msg = "The supergraph should have parameter settings."
+            logging.error(error_msg)
+            raise ValueError(error_msg)
+
+    def __init_supervertex_module_dict__(self):
+        self.supervertex_module_dict: Dict[str, torch.nn.ModuleList] = {}
+        for supervertex_name in self.supergraph.topological_order:
+            supervertex = self.supergraph.supervertex_dict[supervertex_name]
+            setting = self.supergraph.supervertex_setting_dict[supervertex_name]
+            self.__init_module_supervertex__(supervertex, setting)
+
+    def __init_module_supervertex__(self, supervertex: SuperVertex, setting: SuperVertexParaSetting):
+        module_list = torch.nn.ModuleList()
+
+        if not supervertex.if_start_supervertex:
+            # add the external modules from all parent supervertices
+            for in_name in supervertex.in_supervertex_list:
+                in_channels = self.supervertex_module_dict[in_name][-1].out_channels
+
+                assert setting.exter_agg_dim is not None
+                out_channels = setting.exter_agg_dim[in_name]
+                module_list.append(GripNetExternalModule(in_channels, out_channels, supervertex.num_node))
+
+        # add the internal module
+        module_list.append(
+            GripNetInternalModule(
+                supervertex.num_node_feat, supervertex.num_edge_type, supervertex.if_start_supervertex, setting
+            )
+        )
+
+        self.supervertex_module_dict[supervertex.name] = module_list
+
+    def forward(self):
+        for supervertex_name in self.supergraph.topological_order:
+            self.__forward_supervertex__(supervertex_name)
+
+        return self.out_embed_dict[self.task_supervertex_name]
+
+    def __forward_supervertex__(self, supervertex_name: str):
+        assert self.supergraph.supervertex_setting_dict is not None
+
+        supervertex = self.supergraph.supervertex_dict[supervertex_name]
+        model = self.supervertex_module_dict[supervertex_name]
+        mode = self.supergraph.supervertex_setting_dict[supervertex_name].mode
+
+        # internal feature layer
+        x = torch.matmul(supervertex.node_feat, model[-1].embedding)
+
+        # external feature aggregation layers
+        if mode == "add":
+            for idx in range(len(supervertex.in_supervertex_list)):
+                parient_name = supervertex.in_supervertex_list[idx]
+                parient_x = self.out_embed_dict[parient_name]
+                superedge = self.supergraph.superedge_dict[(parient_name, supervertex_name)]
+
+                x += model[idx](parient_x, superedge.edge_index, superedge.edge_weight)
+
+        if mode == "cat":
```

Comment:
will it be better to use elseif (or switch) since there can be only one mode (right?)

Reply:
Yes, changed in [888961a](https://github.com/pykale/pykale/pull/338/commits/888961ae2461a480da9cdc1524498781db773457).